src/utils/allocators.zig

```diff
@@ -587,16 +588,23 @@ pub const DiskMemoryAllocator = struct {
         // the allocator interface shouldn't allow this (aside from the *Raw methods).
         std.debug.assert(alignment <= std.mem.page_size);
 
-        const old_file_aligned_size = alignedFileSize(buf.len);
-        const new_file_aligned_size = alignedFileSize(new_size);
+        const old_file_aligned_size = alignedFileSize(buf.len + @sizeOf(Metadata));
+        const new_file_aligned_size = alignedFileSize(requested_size + @sizeOf(Metadata));
 
         if (new_file_aligned_size == old_file_aligned_size) {
             return true;
         }
 
         const buf_ptr: [*]align(std.mem.page_size) u8 = @alignCast(buf.ptr);
-        const offset = old_file_aligned_size - @sizeOf(Metadata);
-        const metadata: Metadata = @bitCast(buf_ptr[offset..][0..@sizeOf(Metadata)].*);
+        const old_metadata_start = old_file_aligned_size - @sizeOf(Metadata);
+        const metadata: Metadata = @bitCast(blk: {
+            // you might think this block can be replaced with:
+            //      buf_ptr[old_metadata_start..][0..@sizeOf(Metadata)].*
+            // but no, that causes bus errors. it's not the same!
+            var metadata_bytes: [@sizeOf(Metadata)]u8 = undefined;
+            @memcpy(&metadata_bytes, buf_ptr[old_metadata_start..][0..@sizeOf(Metadata)]);
+            break :blk metadata_bytes;
+        });
```

Comment:
Bus errors in this context, on Apple Silicon, are almost always the user's fault, where LLVM expected a certain alignment of the pointer and emitted an instruction with an alignment requirement that wasn't met at runtime.

The issue going away with `@memcpy` makes me pretty sure that's the issue.

This isn't the correct fix, and I'm not sure what it is, since I haven't looked into this myself. Do you have a reproduction for the bus errors?

Reply:
> Do you have a reproduction for the bus errors?

If you replace the code as it says in the comment, it causes a bus error in both macos and linux. Just run `zig build test`.

```patch
diff --git a/src/utils/allocators.zig b/src/utils/allocators.zig
index bd1ef930..6e95dc88 100644
--- a/src/utils/allocators.zig
+++ b/src/utils/allocators.zig
@@ -597,14 +597,7 @@ pub const DiskMemoryAllocator = struct {
 
         const buf_ptr: [*]align(std.mem.page_size) u8 = @alignCast(buf.ptr);
         const old_metadata_start = old_file_aligned_size - @sizeOf(Metadata);
-        const metadata: Metadata = @bitCast(blk: {
-            // you might think this block can be replaced with:
-            //      buf_ptr[old_metadata_start..][0..@sizeOf(Metadata)].*
-            // but no, that causes bus errors. it's not the same!
-            var metadata_bytes: [@sizeOf(Metadata)]u8 = undefined;
-            @memcpy(&metadata_bytes, buf_ptr[old_metadata_start..][0..@sizeOf(Metadata)]);
-            break :blk metadata_bytes;
-        });
+        const metadata: Metadata = @bitCast(buf_ptr[old_metadata_start..][0..@sizeOf(Metadata)].*);
 
         if (new_file_aligned_size > metadata.mmap_size) {
             return false;
```

Reply:
Could it be that the bitcast into the aligned variable is somehow leading zig to memcpy in a misaligned way?
I'm going to try the repro, but with `const metadata: Metadata align(1)`.